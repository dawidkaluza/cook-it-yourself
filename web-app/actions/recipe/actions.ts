"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
