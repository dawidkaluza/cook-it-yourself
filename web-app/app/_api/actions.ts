"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}