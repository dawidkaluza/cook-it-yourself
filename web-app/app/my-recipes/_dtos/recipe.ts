export type Recipe = {
  id?: number;
  name: string;
  description: string;
};

export type Ingredient = {
  id?: number;
  name: string;
  value: string;
  measure: string;
};

export type NewIngredient = Omit<Ingredient, "id">;
export type PersistedIngredient = Ingredient & { id: number };

export type Step = {
  id?: number;
  text: string;
};

export type NewStep = Omit<Step, "id">;
export type PersistedStep = Step & { id: number };

export type PortionSize = {
  value: string;
  measure: string;
};

export type RecipeDetails = Recipe & {
  ingredients: Ingredient[];
  methodSteps: Step[];
  cookingTime: number;
  portionSize: PortionSize;
};

export type UpdateRecipeRequest = {
  basicInformation?: {
    name: string;
    description: string;
    cookingTime: number;
    portionSize: PortionSize;
  };
  ingredients?: {
    ingredientsToAdd: NewIngredient[],
    ingredientsToUpdate: PersistedIngredient[],
    ingredientsToDelete: number[],
  };
  steps?: {
    stepsToAdd: NewStep[],
    stepsToUpdate: PersistedStep[],
    stepsToDelete: number[],
  };
};