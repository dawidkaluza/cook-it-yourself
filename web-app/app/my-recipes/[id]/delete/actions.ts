"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}