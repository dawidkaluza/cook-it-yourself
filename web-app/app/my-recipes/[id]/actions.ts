"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}