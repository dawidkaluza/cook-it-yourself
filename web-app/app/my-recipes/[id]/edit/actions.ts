"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };