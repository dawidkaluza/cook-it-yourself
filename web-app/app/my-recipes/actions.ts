"use server";

import {fetchFromServer} from "@/app/_api/fetch";
import {Page} from "@/app/my-recipes/_dtos/page";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";

export async function getMyRecipes(){
  return await fetchFromServer({ endpoint: "/kitchen/recipe" }) as Promise<Page<Recipe>>;
}