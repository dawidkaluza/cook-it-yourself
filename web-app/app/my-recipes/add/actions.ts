"use server";

import {Ingredient, PortionSize, Recipe, Step} from "@/app/my-recipes/add/_dtos/Recipe";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export type FieldError = {
  name: string;
  message: string;
};

export async function addRecipe(prevState: any, formData: FormData) : Promise<FieldError[]> {
  const mapToIngredients = (formData: FormData) : Ingredient[] => {
    const ingredientsNames  = formData.getAll("ingredientName") as string[];
    const ingredientsAmounts = formData.getAll("ingredientAmount") as string[];
    const ingredientsUnits = formData.getAll("ingredientUnit") as string[];
    const ingredients : Ingredient[] = [];
    const ingredientsNum = ingredientsNames.length;
    for (let i = 0; i < ingredientsNum; i++) {
      const ingredient = {
        name: ingredientsNames[i],
        value: ingredientsAmounts[i],
        measure: ingredientsUnits[i],
      };
      ingredients.push(ingredient);
    }
    return ingredients;
  };

  const mapToSteps = (formData: FormData) : Step[] => {
    return formData.getAll("methodSteps").map(methodStep => {
      return {
        text: methodStep as string,
      };
    });
  };

  const mapToPortionSize = (formData: FormData) : PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  let recipe : Recipe = {
    name: formData.get("name") as string,
    description: formData.get("description") as string,
    ingredients: mapToIngredients(formData),
    methodSteps: mapToSteps(formData),
    cookingTime: Number(formData.get("cookingTime")),
    portionSize: mapToPortionSize(formData)
  };

  try {
    const createdRecipe = await fetchFromServer<Recipe>({
      endpoint: "/kitchen/recipe",
      method: "POST",
      body: recipe,
    });
    if (createdRecipe) {
      redirect(`/my-recipes/${recipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request")
}