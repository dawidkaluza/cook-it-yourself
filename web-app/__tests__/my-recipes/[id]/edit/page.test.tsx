import {useFormState} from "react-dom";
import {Mock} from "vitest";
import {render, screen} from "@testing-library/react";
import {getRecipe} from "@/app/my-recipes/actions";
import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import Page from "@/app/my-recipes/[id]/edit/page";

vi.mock("react-dom", () => {
  return {
    useFormState: vi.fn(),
  };
});

vi.mock("@/app/my-recipes/actions", () => {
  return {
    getRecipe: vi.fn(),
    updateRecipe: vi.fn(),
  };
});

describe("page component", () => {
  test("render", async () => {
    // Given
    const recipeId = 1;
    const recipe: RecipeDetails = {
      id: recipeId,
      name: "Boiled sausages",
      description: "How to boil delicious sausages",
      ingredients: [
        {
          id: 1,
          name: "Sausage",
          value: "3",
          measure: "pc",
        }
      ],
      methodSteps: [
        {
          id: 1,
          text: "Boil sausages for about 3 mins",
        }
      ],
      cookingTime: 3,
      portionSize: {
        value: "1",
        measure: "plate",
      }
    };
    (getRecipe as Mock).mockResolvedValue(recipe);

    (useFormState as Mock).mockReturnValue([
      [],
      () => {}
    ]);

    // When
    const formComponent = render(await Page({ params: { id: recipeId }}));

    // Then
    const nameInput = screen.getByLabelText("Name");
    expect(nameInput).not.toBeNull();

    const descriptionInput = screen.getByLabelText("Description");
    expect(descriptionInput).not.toBeNull();

    const ingredientsInput = screen.getByLabelText("Ingredients");
    expect(ingredientsInput).not.toBeNull();

    const stepsInput = screen.getByLabelText("Method steps");
    expect(stepsInput).not.toBeNull();

    const cookingTimeInput = screen.getByLabelText("Cooking time");
    expect(cookingTimeInput).not.toBeNull();

    const portionSizeInput = screen.getByLabelText("Portion size");
    expect(portionSizeInput).not.toBeNull();

    const submitButton = screen.getByRole("button", { name: /submit/i });
    expect(submitButton).not.toBeNull();

    expect(formComponent.container).toMatchSnapshot();
  });

  test("render with errors", async () => {
    // Given
    const recipeId = 1;
    const recipe: RecipeDetails = {
      id: recipeId,
      name: "Boiled sausages",
      description: "How to boil delicious sausages",
      ingredients: [
        {
          id: 1,
          name: "Sausage",
          value: "3",
          measure: "pc",
        }
      ],
      methodSteps: [
        {
          id: 1,
          text: "Boil sausages for about 3 mins",
        }
      ],
      cookingTime: 3,
      portionSize: {
        value: "1",
        measure: "plate",
      }
    };
    (getRecipe as Mock).mockResolvedValue(recipe);

    const fieldErrors = [
      {
        name: "basicInformation.name",
        message: "Some name error.",
      },
      {
        name: "ingredients.ingredientsToAdd",
        message: "Some ingredients error.",
      }
    ];
    (useFormState as Mock).mockReturnValue([
      fieldErrors,
      () => {}
    ]);

    // When
    const formComponent = render(await Page({ params: { id: recipeId }}));

    // Then
    for (const fieldError of fieldErrors) {
      const errorMessage = screen.getByText(fieldError.message);
      expect(errorMessage).not.toBeNull();
    }
    expect(formComponent.container).toMatchSnapshot();
  });
});