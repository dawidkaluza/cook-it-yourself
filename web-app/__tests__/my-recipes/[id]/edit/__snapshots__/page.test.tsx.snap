// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page component > render 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-12 col-md-10 col-lg-8 col-xl-6"
      >
        <h1>
          Edit the recipe
        </h1>
        <form
          class="mt-4"
          novalidate=""
        >
          <input
            name="id"
            type="hidden"
            value="1"
          />
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="name"
            >
              Name
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="name"
                name="name"
                placeholder="Name of your recipe"
                value="Boiled sausages"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="description"
            >
              Description
            </label>
            <div
              class="col-md-9"
            >
              <textarea
                class="form-control"
                id="description"
                name="description"
                placeholder="A few words describing your recipe"
                style="height: 100px;"
              >
                How to boil delicious sausages
              </textarea>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="newIngredient"
            >
              Ingredients
            </label>
            <div
              class="col-md-9"
            >
              <div
                class="row"
              />
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    name="ingredientId"
                    type="hidden"
                    value="1"
                  />
                  <input
                    class="form-control"
                    name="ingredientName"
                    placeholder="Name"
                    style="min-width: 50%;"
                    value="Sausage"
                  />
                  <input
                    class="form-control"
                    name="ingredientValue"
                    placeholder="Amount"
                    style="min-width: 15%;"
                    value="3"
                  />
                  <input
                    class="form-control"
                    name="ingredientMeasure"
                    placeholder="Unit"
                    style="min-width: 15%;"
                    value="pc"
                  />
                  <button
                    class="btn btn-outline-danger"
                    type="button"
                  >
                    <svg
                      class="bi bi-dash-lg"
                      fill="currentColor"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    class="form-control"
                    id="newIngredient"
                    name="newIngredient"
                    placeholder="e.g., water, apple 2, sugar 250g"
                    value=""
                  />
                  <button
                    class="btn btn-success"
                    type="button"
                  >
                    <svg
                      class="bi bi-plus-lg"
                      fill="currentColor"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="row"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="newMethodStep"
            >
              Method steps
            </label>
            <div
              class="col-md-9"
            >
              <div
                class="row"
              />
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    name="methodStepId"
                    type="hidden"
                    value="1"
                  />
                  <textarea
                    class="form-control"
                    name="methodStepText"
                    placeholder="Method step"
                    style="height: 100px;"
                  >
                    Boil sausages for about 3 mins
                  </textarea>
                  <div
                    class="input-group-text"
                  >
                    <button
                      class="btn btn-sm btn-outline-danger"
                      type="button"
                    >
                      <svg
                        class="bi bi-dash-lg"
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <textarea
                    class="form-control"
                    id="newMethodStep"
                    name="newMethodStep"
                    placeholder="Instructions step-by-step to cook your recipe"
                    style="height: 100px;"
                  />
                  <div
                    class="input-group-text"
                  >
                    <button
                      class="btn btn-sm btn-success"
                      type="button"
                    >
                      <svg
                        class="bi bi-plus-lg"
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="row"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="cookingTime"
            >
              Cooking time
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="cookingTime"
                name="cookingTime"
                placeholder="In minutes"
                type="number"
                value="3"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="portionSize"
            >
              Portion size
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="portionSize"
                name="portionSize"
                placeholder="Amount and unit (e.g., 4 plates, 800g)"
                value="1 plate"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <div
              class="col-12 d-flex justify-content-center"
            >
              <div
                class="btn-group"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  <svg
                    class="bi bi-check2"
                    fill="currentColor"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
                    />
                  </svg>
                   Submit 
                </button>
                <a
                  class="btn btn-outline-secondary"
                  href="/my-recipes/1"
                >
                  Cancel
                </a>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`page component > render with errors 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-12 col-md-10 col-lg-8 col-xl-6"
      >
        <h1>
          Edit the recipe
        </h1>
        <form
          class="mt-4"
          novalidate=""
        >
          <input
            name="id"
            type="hidden"
            value="1"
          />
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="name"
            >
              Name
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="name"
                name="name"
                placeholder="Name of your recipe"
                value="Boiled sausages"
              />
              <div
                class="invalid-feedback d-block"
              >
                Some name error.
              </div>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="description"
            >
              Description
            </label>
            <div
              class="col-md-9"
            >
              <textarea
                class="form-control"
                id="description"
                name="description"
                placeholder="A few words describing your recipe"
                style="height: 100px;"
              >
                How to boil delicious sausages
              </textarea>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="newIngredient"
            >
              Ingredients
            </label>
            <div
              class="col-md-9"
            >
              <div
                class="row"
              />
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    name="ingredientId"
                    type="hidden"
                    value="1"
                  />
                  <input
                    class="form-control"
                    name="ingredientName"
                    placeholder="Name"
                    style="min-width: 50%;"
                    value="Sausage"
                  />
                  <input
                    class="form-control"
                    name="ingredientValue"
                    placeholder="Amount"
                    style="min-width: 15%;"
                    value="3"
                  />
                  <input
                    class="form-control"
                    name="ingredientMeasure"
                    placeholder="Unit"
                    style="min-width: 15%;"
                    value="pc"
                  />
                  <button
                    class="btn btn-outline-danger"
                    type="button"
                  >
                    <svg
                      class="bi bi-dash-lg"
                      fill="currentColor"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    class="form-control"
                    id="newIngredient"
                    name="newIngredient"
                    placeholder="e.g., water, apple 2, sugar 250g"
                    value=""
                  />
                  <button
                    class="btn btn-success"
                    type="button"
                  >
                    <svg
                      class="bi bi-plus-lg"
                      fill="currentColor"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="invalid-feedback d-block"
                >
                  Some ingredients error.
                </div>
              </div>
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="newMethodStep"
            >
              Method steps
            </label>
            <div
              class="col-md-9"
            >
              <div
                class="row"
              />
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <input
                    name="methodStepId"
                    type="hidden"
                    value="1"
                  />
                  <textarea
                    class="form-control"
                    name="methodStepText"
                    placeholder="Method step"
                    style="height: 100px;"
                  >
                    Boil sausages for about 3 mins
                  </textarea>
                  <div
                    class="input-group-text"
                  >
                    <button
                      class="btn btn-sm btn-outline-danger"
                      type="button"
                    >
                      <svg
                        class="bi bi-dash-lg"
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M2 8a.5.5 0 0 1 .5-.5h11a.5.5 0 0 1 0 1h-11A.5.5 0 0 1 2 8"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="input-group"
                >
                  <textarea
                    class="form-control"
                    id="newMethodStep"
                    name="newMethodStep"
                    placeholder="Instructions step-by-step to cook your recipe"
                    style="height: 100px;"
                  />
                  <div
                    class="input-group-text"
                  >
                    <button
                      class="btn btn-sm btn-success"
                      type="button"
                    >
                      <svg
                        class="bi bi-plus-lg"
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="row"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="cookingTime"
            >
              Cooking time
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="cookingTime"
                name="cookingTime"
                placeholder="In minutes"
                type="number"
                value="3"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <label
              class="col-md-3 col-form-label text-md-end"
              for="portionSize"
            >
              Portion size
            </label>
            <div
              class="col-md-9"
            >
              <input
                class="form-control"
                id="portionSize"
                name="portionSize"
                placeholder="Amount and unit (e.g., 4 plates, 800g)"
                value="1 plate"
              />
            </div>
          </div>
          <div
            class="row mb-4"
          >
            <div
              class="col-12 d-flex justify-content-center"
            >
              <div
                class="btn-group"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  <svg
                    class="bi bi-check2"
                    fill="currentColor"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
                    />
                  </svg>
                   Submit 
                </button>
                <a
                  class="btn btn-outline-secondary"
                  href="/my-recipes/1"
                >
                  Cancel
                </a>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
